File: ios/Classes/Bridges/SuperwallDelegateProxyBridge.swift
import Flutter
import SuperwallKit

public class SuperwallDelegateProxyBridge: BridgeInstance, SuperwallDelegate {
  override class var bridgeClass: BridgeClass { "SuperwallDelegateProxyBridge" }

  // MARK: - SuperwallDelegate

  public func willPresentPaywall(withInfo paywallInfo: PaywallInfo) {
    communicator.invokeMethodOnMain("willPresentPaywall", arguments: ["paywallInfoBridgeId": paywallInfo.createBridgeId()])
  }

  public func didPresentPaywall(withInfo paywallInfo: PaywallInfo) {
    communicator.invokeMethodOnMain("didPresentPaywall", arguments: ["paywallInfoBridgeId": paywallInfo.createBridgeId()])
  }

  public func willDismissPaywall(withInfo paywallInfo: PaywallInfo) {
    communicator.invokeMethodOnMain("willDismissPaywall", arguments: ["paywallInfoBridgeId": paywallInfo.createBridgeId()])
  }

  public func didDismissPaywall(withInfo paywallInfo: PaywallInfo) {
    communicator.invokeMethodOnMain("didDismissPaywall", arguments: ["paywallInfoBridgeId": paywallInfo.createBridgeId()])
  }

  public func handleCustomPaywallAction(withName name: String) {
    communicator.invokeMethodOnMain("handleCustomPaywallAction", arguments: ["name": name])
  }

  public func subscriptionStatusDidChange(to newValue: SubscriptionStatus) {
    communicator.invokeMethodOnMain("subscriptionStatusDidChange", arguments: ["subscriptionStatusBridgeId": newValue.createBridgeId()])
  }

  public func handleSuperwallEvent(withInfo eventInfo: SuperwallEventInfo) {
    communicator.invokeMethodOnMain("handleSuperwallEvent", arguments: ["eventInfo": eventInfo.toJson()])
  }

  public func paywallWillOpenURL(url: URL) {
    communicator.invokeMethodOnMain("paywallWillOpenURL", arguments: ["url" : url.absoluteString])
  }

  public func paywallWillOpenDeepLink(url: URL) {
    communicator.invokeMethodOnMain("paywallWillOpenDeepLink", arguments: ["url" : url.absoluteString])
  }

  public func handleLog(
    level: String,
    scope: String,
    message: String?,
    info: [String: Any]?,
    error: Swift.Error?
  ) {
    let transformedInfo = info?.compactMapValues { value -> Any? in
      switch value {
        case let stringValue as String:
          return stringValue
        case let intValue as Int:
          return intValue
        case let dictValue as [String: Any]:
          return dictValue
        case let arrayValue as [Any]:
          return arrayValue
        case let boolValue as Bool:
          return boolValue
        case let setValue as Set<AnyHashable>:
          return Array(setValue)
        default:
          return nil
      }
    }

    let arguments: [String: Any?] = [
      "level": level,
      "scope": scope,
      "message": message,
      "info": transformedInfo,
      "error": error?.localizedDescription
    ]
    communicator.invokeMethodOnMain("handleLog", arguments: arguments)
  }
}

File: android/src/main/kotlin/com/superwall/superwallkit_flutter/bridges/SuperwallDelegateProxyBridge.kt
package com.superwall.superwallkit_flutter.bridges

import android.content.Context
import android.net.Uri
import com.superwall.sdk.analytics.superwall.SuperwallPlacementInfo
import com.superwall.sdk.models.entitlements.SubscriptionStatus
import com.superwall.sdk.delegate.SuperwallDelegate
import com.superwall.sdk.paywall.presentation.PaywallInfo
import com.superwall.superwallkit_flutter.invokeMethodOnMain
import toJson
import java.net.URI
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class SuperwallDelegateProxyBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>? = null,
    val coroutineScope: CoroutineScope = CoroutineScope(Dispatchers.IO)
) : BridgeInstance(context, bridgeId, initializationArgs), SuperwallDelegate {
    companion object {
        fun bridgeClass(): BridgeClass = "SuperwallDelegateProxyBridge"
    }

    override fun willPresentPaywall(paywallInfo: PaywallInfo) {
        coroutineScope.launch {
            communicator().invokeMethodOnMain(
                "willPresentPaywall",
                mapOf("paywallInfoBridgeId" to paywallInfo.createBridgeId())
            )
        }
    }

    override fun didPresentPaywall(paywallInfo: PaywallInfo) {
        coroutineScope.launch {

            communicator().invokeMethodOnMain(
                "didPresentPaywall",
                mapOf("paywallInfoBridgeId" to paywallInfo.createBridgeId())
            )
        }
    }

    override fun willDismissPaywall(paywallInfo: PaywallInfo) {
        coroutineScope.launch {

            communicator().invokeMethodOnMain(
                "willDismissPaywall",
                mapOf("paywallInfoBridgeId" to paywallInfo.createBridgeId())
            )
        }
    }

    override fun didDismissPaywall(paywallInfo: PaywallInfo) {
        coroutineScope.launch {
            communicator().invokeMethodOnMain(
                "didDismissPaywall",
                mapOf("paywallInfoBridgeId" to paywallInfo.createBridgeId())
            )
        }
    }

    override fun handleCustomPaywallAction(name: String) {
        coroutineScope.launch {
            communicator().invokeMethodOnMain("handleCustomPaywallAction", mapOf("name" to name))
        }
    }

    override fun subscriptionStatusDidChange(from: SubscriptionStatus, to: SubscriptionStatus) {
        coroutineScope.launch {
            communicator().invokeMethodOnMain(
                "subscriptionStatusDidChange",
                mapOf("subscriptionStatusBridgeId" to to.createBridgeId())
            )
        }
    }

    override fun handleSuperwallPlacement(placementInfo: SuperwallPlacementInfo) {
        coroutineScope.launch {
            communicator().invokeMethodOnMain(
                "handleSuperwallPlacement",
                mapOf("placementInfo" to placementInfo.toJson())
            )
        }
    }

    override fun paywallWillOpenURL(url: URI) {
        coroutineScope.launch {
            communicator().invokeMethodOnMain("paywallWillOpenURL", mapOf("url" to url.toString()))
        }
    }

    override fun paywallWillOpenDeepLink(url: Uri) {
        coroutineScope.launch {
            communicator().invokeMethodOnMain(
                "paywallWillOpenDeepLink",
                mapOf("url" to url.toString())
            )
        }
    }

    override fun handleLog(
        level: String,
        scope: String,
        message: String?,
        info: Map<String, Any>?,
        error: Throwable?
    ) {
        coroutineScope.launch {
            val transformedInfo = info?.mapValues { (_, value) ->
                when (value) {
                    is String -> value
                    is Int -> value
                    is Map<*, *> -> value
                    is List<*> -> value
                    is Boolean -> value
                    is Set<*> -> value.toList()
                    else -> null
                }
            }

            val arguments: Map<String, Any?> = mapOf(
                "level" to level,
                "scope" to scope,
                "message" to message,
                "info" to transformedInfo,
                "error" to error?.localizedMessage
            )

            communicator().invokeMethodOnMain("handleLog", arguments)
        }
    }
}

